tests/rest/client/test_room_batch.py

```diff
@@ -178,3 +179,112 @@ def test_same_state_groups_for_whole_historical_batch(self) -> None:
             "Expected a single state_group to be returned by saw state_groups=%s"
             % (state_group_map.keys(),),
         )
+
+    @unittest.override_config({"experimental_features": {"msc2716_enabled": True}})
+    def test_sync_while_batch_importing(self) -> None:
+        """
+        Make sure that /sync correctly returns full room state when a user joins
+        during ongoing batch backfilling.
+        See: https://github.com/matrix-org/synapse/issues/12281
+        """
+        # Create user who will be invited & join room
+        user_id = self.register_user("beep", "test")
+        user_tok = self.login("beep", "test")
+
+        time_before_room = int(self.clock.time_msec())
+
+        # Create a room with some events
+        room_id, _, _, _ = self._create_test_room()
+        # Invite the user
+        self.helper.invite(
+            room_id, src=self.appservice.sender, tok=self.appservice.token, targ=user_id
+        )
+
+        # Create another room, send a bunch of events to advance the stream token
+        other_room_id = self.helper.create_room_as(
+            self.appservice.sender, tok=self.appservice.token
+        )
+        for _ in range(5):
+            self.helper.send_event(
+                room_id=other_room_id,
+                type=EventTypes.Message,
+                content={"msgtype": "m.text", "body": "C"},
+                tok=self.appservice.token,
+            )
+
+        # Join the room as the normal user
+        self.helper.join(room_id, user_id, tok=user_tok)
+
+        # Create event to hang backfill batch from, and send batch
+        response = self.helper.send_event(
+            room_id=room_id,
+            type=EventTypes.Message,
+            content={
+                "msgtype": "m.text",
+                "body": "C",
+            },
+            tok=self.appservice.token,
+        )
+        event_to_hang_id = response["event_id"]
```

Comment:
Instead of creating a new `event_to_hang_id`. We can use one of the events that `_create_test_room` already creates with the room

```py
room_id, event_id_a, _, _ = self._create_test_room()
```

Reply:
ehhh, nvm. It's important that we `/batch_send` off of the latest event in the room. 

But we should add a comment about why this is important ^

Reply:
https://github.com/matrix-org/synapse/pull/12319/commits/b2c6c20f4ab8e68b56b39b869af29c56e81d0ff7

Reply:
Review continued here https://github.com/matrix-org/synapse/pull/12319/files#r848686533